### ISSUE TITLE
Errors on Ubuntu

### ISSUE THREAD

**Issue by Alexander-Shukaev** [abingham/emacs-ycmd#232]
Have to work on a different machine now.  It has Ubuntu 14 installed.  I have successfully built YCM as in the official guide, and, as a result, have now this `ycmd` directory.  However, I keep receiving periodically:

```
REQUEST [error] Error (error) while connecting to http://127.0.0.1:52426/healthy.
```

and accordingly when I try, for instnace, `ycmd-goto`, I get

```
REQUEST [error] Error (parse-error) while connecting to http://127.0.0.1:52426/run_completer_command.
```

On my home Arch Linux, I have no problems so far.  So what should I look for in this case?  Thanks in advance.

**Comment by abingham** [comment]
Do you see anything suspicious-looking in the `*ycmd-server*` buffer?

**Comment by Alexander-Shukaev** [comment]
I currently see the following:

```
2015-11-17 10:51:31,862 - DEBUG - Global extra conf not loaded or no function YcmCorePreload
serving on http://127.0.0.1:52426
```

**Comment by abingham** [comment]
Hmm. What if you point your browser to http://127.0.0.1:52426/healthy (or whatever port your ycmd server has)? I'm trying to figure out if it can be reached at all.

**Comment by Alexander-Shukaev** [comment]
Says

```
Error: 401 Unauthorized

Sorry, the requested URL 'http://127.0.0.1:52426/healthy' caused an error:

Unauthorized, received bad HMAC.
```

**Comment by abingham** [comment]
OK, that's good, at least it's responding to HTTP. The next thing to try is to enable HTTP logging and see if there are any clues there. 

Enable logging:
```
(setq ycmd--log-enabled t)
``` 
Then try a `ycmd-goto` or anything else and then disable logging:
```
(setq ycmd--log-enabled nil)
```
The logging output should be in `*ycmd-content-log*`.

**Comment by Alexander-Shukaev** [comment]
There it goes:

```
HTTP REQUEST CONTENT

"{\"command_arguments\":[\"GoTo\"],\"file_data\":{\"...\":{\"contents\":\"...\",\"filetypes\":[\"cpp\"]}},\"filepath\":\"...\",\"line_num\":147,\"column_num\":13}"
HTTP RESPONSE CONTENT

nil
HTTP REQUEST CONTENT

"null"
HTTP RESPONSE CONTENT

"<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01 Transitional//EN\" \"http://www.w3.org/TR/html4/loose.dtd\">\n<html>\n  <head>\n    <meta http-equiv=\"Content-Type\" content=\"text/html; charset=UTF-8\" />\n    <title>The requested URL could not be retrieved</title>\n    <link href=\"http://passthrough.fw-notify.net/static/default.css\" rel=\"stylesheet\" type=\"text/css\" />\n    <script type=\"text/javascript\" src=\"http://passthrough.fw-notify.net/static/default.js\"></script>\n  </head>\n  <body onload=\"checkResize();\">\n    <div id=\"emsg_large\"></div>\n    <div id=\"page\">\n      <div id=\"header\">\n        <div><img src=\"http://passthrough.fw-notify.net/static/topbar_left.png\" width=\"6\" height=\"72\" border=\"0\" /></div>\n        <div id=\"company_logo\"><img src=\"http://passthrough.fw-notify.net/static/logo.png\" border=\"0\" /></div>\n        <div id=\"company_text\"><h1 class=\"orange\">Sophos UTM Version 9   http://www.sophos.com</h1></div>\n        <div><img src=\"http://passthrough.fw-notify.net/static/topbar_right.png\" width=\"6\" height=\"72\" border=\"0\" /></div>\n      </div>\n      <br class=\"clearer\" />\n\n      <div id=\"content\">\n        <img src=\"http://passthrough.fw-notify.net/static/warning.png\" border=\"0\" align=\"left\" />\n        <h1 class=\"orange\">An error occurred while handling your request</h1>\n\n        <div class=\"line\">\n          <div class=\"label\">While trying to retrieve the URL:</div>\n          <div class=\"desc\"> \n            <span>http://127.0.0.1:47048/healthy</span>\n          </div>\n        </div>\n\n        <div class=\"line\">\n          <div class=\"label\">The content could not be delivered due to the following condition:</div>\n          <div class=\"desc\"> \n            Target service not allowed\n          </div>\n        </div>\n\n        <div class=\"line\">\n        </div>\n        <br class=\"clearer\" />\n      </div> <!-- END: div#content -->\n      <br class=\"clearer\" />\n\n      <div id=\"footer\">\n        <div><a href=\"http://www.sophos.com\" target=\"_blank\" title=\"Sophos\"><img src=\"http://passthrough.fw-notify.net/static/footer_left.png\" width=\"136\" height=\"33\" border=\"0\" /></a></div>\n        <div class=\"orange\" id=\"copyright\">Powered by UTM Web Protection</div>\n        <div><img src=\"http://passthrough.fw-notify.net/static/footer_right.png\" width=\"6\" height=\"33\" border=\"0\" /></div>\n      </div> <!-- END: div#footer -->\n    </div> <!-- END: div#page -->\n  </body>\n</html>\n"
```

**NOTE:**  I've restarted the server, so the port is different now.  I have omitted source code and paths because this is irrelevant anyway.

**Comment by abingham** [comment]
It looks like you've got a Sophos firewall blocking the service. That blob
of HTML you posted mentions "Sophos UTM Version 9" and says "The content
could not be delivered due to the following condition: Target service not
allowed". You should see about disabling that.

**Comment by Alexander-Shukaev** [comment]
Alright, so after asking some experts, I was told that as long as YCMD runs locally and sends requests locally (does not involve any network HTTP traffic), then it cannot be blocked by Sophos.  I was then suggested to simply unplug the network cable to confirm that point as I was advocating that YCMD is not supposed to have any network HTTP traffic.  When I unplugged the network cable, to my surprise either `ycmd-open` or opening some C/C++ source file leads to a dead freeze of Emacs, that is I can only recover by killing the Emacs process.  Any ideas what's going on?

**Comment by abingham** [comment]
You're right that, as far as I know, ycmd should not be involved with any HTTP traffic outside of the localhost. I suppose it's possible that you've got some sort of http proxy that forces all traffic out of the machine or something...I'm not an expert in that area. 

One quick way to maybe test this is to run an HTTP server on one of the ports you've had trouble with and see what happens. If you've got Python3, something like this should work:
```
echo "hello" > test.html
python3 -m http.server 52426
```
Now if you point a browser at localhost:52426/test.html, if there's some sort of firewall or proxy affecting all HTTP traffic on your machine you should see that warning from sophos. It's worth a shot to try to narrow down the problem space...

**Comment by abingham** [comment]
As for the freezing issue, could you make a new issue for that and link it to this one? That's really odd, and it *does* seem to indicate that HTTP traffic to your ycmd is going through the cable.

**Comment by Alexander-Shukaev** [comment]
Just tried your suggestion and I can successfully load the page.  No Sophos messages are present.

**Comment by Alexander-Shukaev** [comment]
One detail though, it says:

```
Serving HTTP on 0.0.0.0 port 52426 ...
```

So it's `0.0.0.0` rather than `127.0.0.1` as is the case with YCMD.

**Comment by abingham** [comment]
Try `python3 -m http.server --bind 127.0.0.1 52426`, just for completeness.

**Comment by Alexander-Shukaev** [comment]
Yep, works as well.

**Comment by abingham** [comment]
Hmmm...ok. If you don't mind, I'd like you to test the connectivity between emacs and ycmd. Here's a bit of elisp code that will make an HTTP request and dump the data into the `*Messages*` buffer:
```
(ycmd-request
 "http://127.0.0.1:43567/test.html"
 :parser 'buffer-string
 :success
 (function* (lambda (&key data &allow-other-keys)
              (message "RESPONSE DATA: %s" data))))
```
Modify it to match your system and let me know how it behaves.

**Comment by Alexander-Shukaev** [comment]
I found out what was the problem.  So, since we have HTTP proxies installed here, one has the `http_proxy` environment variable set.  Hence, Emacs sets `url-proxy-services` automatically:

```
url-proxy-services is a variable defined in `url-vars.el'.
Its value is (("http" . "$http_proxy:80"))
Original value was nil

Documentation:
An alist of schemes and proxy servers that gateway them.
Looks like (("http" . "hostname:portnumber") ...).  This is set up
from the ACCESS_proxy environment variables.

You can customize this variable.
```

This is also necessary for `package.el` to be able to download packages, for example.  So what I did was:

```
(add-hook 'after-init-hook #'(lambda ()
                               (setq-default url-proxy-services nil)))
```

And it YCMD works now as expected.  My suggestion would be to add something like `ycmd-url-proxy-services` which can override `url-proxy-services` with `let` binding around the important connection calls inside `ycmd.el`.  The default value would obviously be `nil` as running YCMD locally is the most common scenario.  In any case, thanks for your help and effort!

**Comment by abingham** [comment]
Excellent news! I'm glad you were able to sort this out, and thank you for putting in the energy to resolve it. I've created a new issue to implement what you've described. Hopefully we'll get to it soon.

If you're satisfied with your current fix, then go ahead and close this out.

**Comment by Alexander-Shukaev** [comment]
Yep, thanks!
